src/ensembl/src/shared/select/SelectOptionsPanel.tsx

```diff
@@ -0,0 +1,307 @@
+
+import styles from './Select.scss';
+
+enum HighlightActionType {
+  NEXT = 'next',
+  PREVIOUS = 'previous',
+  SET = 'set'
+}
+type HighlightedItemState = GroupedOptionIndex | null;
+type HighlightAction =
+  | { type: HighlightActionType.NEXT; payload: OptionGroup[] }
+  | { type: HighlightActionType.PREVIOUS; payload: OptionGroup[] }
+  | { type: HighlightActionType.SET; payload: GroupedOptionIndex };
+
+type OptionGroupProps = OptionGroup & {
+  highlightedItemIndex?: number;
+  groupIndex: number;
+  onItemHover: (index: GroupedOptionIndex) => void;
+  onItemClick: (index: GroupedOptionIndex) => void;
+};
+
+type OptionProps = Option & {
+  groupIndex: number;
+  itemIndex: number;
+  isHighlighed: boolean;
+  onHover: (index: GroupedOptionIndex) => void;
+  onClick: (index: GroupedOptionIndex) => void;
+};
+
+type Props = {
+  optionGroups: OptionGroup[];
+  header: React.ReactNode;
+  onSelect: (index: GroupedOptionIndex) => void;
+  onClose: () => void;
+};
+
+const highlightedItemReducer = (
+  state: HighlightedItemState,
+  action: HighlightAction
+) => {
+  switch (action.type) {
+    case HighlightActionType.NEXT:
+      return getNextItemIndex(state, action.payload);
+    case HighlightActionType.PREVIOUS:
+      return getPreviousItemIndex(state, action.payload);
+    case HighlightActionType.SET:
+      return action.payload;
+    default:
+      return state;
+  }
+};
+
+const SelectOptionsPanel = (props: Props) => {
+  const [shouldShowTopScrollButton, setShouldShowTopScrollButton] = useState(
+    false
+  );
+  const [
+    shouldShowBottomScrollButton,
+    setShouldShowBottomScrollButton
+  ] = useState(false);
+  const [highlightedItemIndex, dispatch] = useReducer(
+    highlightedItemReducer,
+    null
+  );
+  const elementRef = useRef<HTMLDivElement | null>(null);
+  const optionsListRef = useRef<HTMLDivElement | null>(null);
+  const scrollRef = useRef(0);
+
+  // to be able to reach the value of highlighted index
+  // from within callbacks passed to useEffect
+  const highlightedItemIndexRef = useRef(highlightedItemIndex);
+  highlightedItemIndexRef.current = highlightedItemIndex;
+
+  const handleKeyDown = (event: KeyboardEvent) => {
+    if (![keyCodes.UP, keyCodes.DOWN, keyCodes.ENTER].includes(event.keyCode)) {
+      return;
+    }
+
+    event.preventDefault();
+
+    if (event.keyCode === keyCodes.UP) {
+      dispatch({
+        type: HighlightActionType.PREVIOUS,
+        payload: props.optionGroups
+      });
+    } else if (event.keyCode === keyCodes.DOWN) {
+      dispatch({ type: HighlightActionType.NEXT, payload: props.optionGroups });
+    } else if (event.keyCode === keyCodes.ENTER) {
+      highlightedItemIndexRef.current &&
+        props.onSelect(highlightedItemIndexRef.current);
+    }
+
+    // scroll option into view after the state gets updated
+    setTimeout(() => {
+      if (!optionsListRef.current) {
+        // keypress has caused the options panel to close
```

Comment:
Does this block gets executed only when `enter` key is pressed? If yes, I think we could use a `return` within the previous `if (event.keyCode === keyCodes.ENTER) {..` block and remove this check.

Reply:
This is a Typescript thing. `optionsListRef` is defined above as `HTMLDivElement | null`. If we move the `return` statement into the `Enter` branch, we will still need to convince Typescript that `optionsListRef.current` is not empty. We can do that, of course, by telling Typescript that we know better than it (`const optionsListElement = optionsList.current as  HTMLDivElement`), but at the moment Typescript, together with the if-statement, are doing this for us, and are safer than any manual override.